### ISSUE TITLE
Rework IEntity interface and entity definition

### ISSUE THREAD

**Issue by thynson** [mikro-orm/mikro-orm#171]
I'm invesgating this package, and very satisfied with the promising features.
However, it make me a little uncomfortable that entity class need to extends `IEntity` interface, as it just pollute the namespace.
```typescript

@Entity()
class MyEntity {

}

// isInitialized added to MyEntity as it's a member function of IEntity
// while it's a very common name it may be overrided accidently
interface MyEntity extends IEntity {}

// isInitialized is typed to be always presents, but actually it's not
new MyEntity().isInitialized(); // TypeError: (intermediate value).isInitialized is not a function
``` 
I guess that entity manager need need to track instance through the help of `IEntity`, but this could be in a better way, e.g. using symbol instead.

**Comment by B4nan** [comment]
Yeah I am not very happy with that. How would using symbol help here? The main reason for this is to add that extra public API to all entities (like `init` or `isInitialized`). 

> new MyEntity().isInitialized(); // TypeError: (intermediate value).isInitialized is not a function

This code should work if you started the ORM first (aka the discovery process went thru this entity).

**Comment by thynson** [comment]
I suppose fields prefixed with `__` like `__uuid` is not expected to be used out use code. 
```typescript
// Entity.ts
const _UUIDKey = Symbol(); 

export interface MyInterface {
    [_UUIDKey]: string; // this field can only be access in this file as _UUIDKey is not exported
}

// SomeEntity.ts

export class SomeEntity {

}
export interface MyEntity extends MyInterface {}
```

For some extra public API like `isInitialized()`, I think they shoud be repository method to make entity class clean and  persistence ignorance.

**Comment by B4nan** [comment]
I was also thinking about making all the methods defined as optional, obviously then to use them you would have to add `!`:

```typescript
if (await entity.isInitialized!()) {
  await entity.init!();
}
```

With that approach it should be possible to simply implement the interface (as it would contain only optional fields and methods):

```typescript
export class MyEntity implements IEntity { ... }
```

> I suppose fields prefixed with __ like __uuid is not expected to be used out use code.

Yeah I was going to wrap those into `__internal`, we can have it as a symbol, but as it is prefixed with `__` I do not think we need to care much about collisions... Also I would have to export the symbol as there are multiple places where it is used.

> For some extra public API like isInitialized(), I think they shoud be repository method to make entity class clean and persistence ignorance.

I think it should be ability of the entity to tell whether it is loaded or not, the information is there already so why involving EM. I want users to be able to work with entities in services without requiring EM reference. Same applies to other methods, like `toJSON` and `toObject`, having that on EM/repository level seems clunky.

**Comment by thynson** [comment]
> I think it should be ability of the entity to tell whether it is loaded or not, the information is there already so why involving EM. I want users to be able to work with entities in services without requiring EM reference. Same applies to other methods, like toJSON and toObject, having that on EM/repository level seems clunky.

If entity could have `isInitialized` method, they should have `save`, `update`, etc, and then we get a ActiveRecord style ORM. (while `toJSON` and `toObject` ar  not related to persistence. so I'm fine with them)

TypeORM provides both ActiveRecord style and EntityFramework style. I just prefer latter. :)

**Comment by B4nan** [comment]
> If entity could have isInitialized method, they should have save, update, etc, and then we get a ActiveRecord style ORM.

I don't agree here, this is data mapper, having one method on entity interface does not make it active record. You have to work with entity manager to persist things. This whole `init/isInitialized` API is there because entity can be just a reference (not yet populated entity that has only PK). `init` is there just to make things easier from developer's perspective, it just does `this.__em.findOne(this.id)` in the background. 

I recently added `Reference` wrapper that could end up being a solution to this, if we always use it for references. But that would break comparing entities by identity probably. Currently its usage is optional. 

https://mikro-orm.github.io/docs-v3/entity-references/

**Comment by B4nan** [comment]
Another idea: keep IEntity interface without methods and props so it can be implemented instead of merged, and add a helper method that would return wrapped entity with all those new methods like init etc:

```typescript
import { wrap } from mikro-orm;

await wrap(myEntity).init();
```

I would probably keep those methods there as optional, so one could also do this: 

```typescript
await myEntity.init!();
```

**Comment by thynson** [comment]
I prefer the wrap style to avoid polluting the interface.

**Comment by B4nan** [comment]
Ok, after battling with #124 I see this is really needed. I will go for `wrap` helper as described above. This is how normal entity (sql driver, AI number PK) will look like:

```typescript
@Entity()
export class FooBar implements IdentifiedEntity<FooBar> {

  @PrimaryKey()
  id: number;

  @Property()
  name: string;

}
```

For mongo it will be a bit more complicated as now both `_id` and `id` will need to be defined: 

```typescript
@Entity()
export class FooBar implements MongoEntity<FooBar> {

  @PrimaryKey()
  _id: ObjectId;

  @SerializedPrimaryKey()
  id: string;

  @Property()
  name: string;

}
```

There will be also `UuidEntity` interface where PK will be defined as `uuid: string`, and also `IEntity` interface where one can define any property as PK (defining PK property will be needed for `FilterQuery` strict typing).

**Comment by B4nan** [comment]
Available in `3.0.0-alpha.23`
